File: app/src/main/java/orinasa/njarasoa/maripanatokana/data/remote/OpenMeteoMapper.kt
package orinasa.njarasoa.maripanatokana.data.remote

import orinasa.njarasoa.maripanatokana.domain.model.DailyForecast
import orinasa.njarasoa.maripanatokana.domain.model.HourlyForecast
import orinasa.njarasoa.maripanatokana.domain.model.Precipitation
import orinasa.njarasoa.maripanatokana.domain.model.Pressure
import orinasa.njarasoa.maripanatokana.domain.model.Temperature
import orinasa.njarasoa.maripanatokana.domain.model.WeatherData
import orinasa.njarasoa.maripanatokana.domain.model.WindSpeed
import java.util.Calendar
import java.util.TimeZone

internal fun parseIsoDateTime(iso: String, utcOffsetSeconds: Int): Long {
    try {
        val year = iso.substring(0, 4).toInt()
        val month = iso.substring(5, 7).toInt() - 1
        val day = iso.substring(8, 10).toInt()
        val hour = iso.substring(11, 13).toInt()
        val minute = iso.substring(14, 16).toInt()
        val calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC")).apply {
            set(year, month, day, hour, minute, 0)
            set(Calendar.MILLISECOND, 0)
        }
        return calendar.timeInMillis - (utcOffsetSeconds * 1000L)
    } catch (e: Exception) {
        return 0L
    }
}

private fun parseIsoDate(iso: String, utcOffsetSeconds: Int): Long {
    try {
        val year = iso.substring(0, 4).toInt()
        val month = iso.substring(5, 7).toInt() - 1
        val day = iso.substring(8, 10).toInt()
        val calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC")).apply {
            set(year, month, day, 0, 0, 0)
            set(Calendar.MILLISECOND, 0)
        }
        return calendar.timeInMillis - (utcOffsetSeconds * 1000L)
    } catch (e: Exception) {
        return 0L
    }
}


fun OpenMeteoResponse.toDomain(locationName: String, locationSubtext: String? = null): WeatherData {
    val c = current
    val isDay = c.isDay == 1

    // Pre-parse hourly timestamps once
    val parsedHourlyTimes = hourly.time.map { parseIsoDateTime(it, utcOffsetSeconds) }

    val sunriseEpoch = daily.sunrise.firstOrNull()?.let {
        parseIsoDateTime(it, utcOffsetSeconds) / 1000
    } ?: 0L
    val sunsetEpoch = daily.sunset.firstOrNull()?.let {
        parseIsoDateTime(it, utcOffsetSeconds) / 1000
    } ?: 0L

    val dailySunriseMillis = daily.sunrise.map { parseIsoDateTime(it, utcOffsetSeconds) }
    val dailySunsetMillis = daily.sunset.map { parseIsoDateTime(it, utcOffsetSeconds) }

    val nowMillis = System.currentTimeMillis()

    // Find the first index using correctly parsed absolute UTC times
    val startIndex = parsedHourlyTimes.indexOfFirst { it >= nowMillis }.takeIf { it != -1 } ?: 0

    val endIndex = minOf(startIndex + 24, hourly.time.size)

    val hourlyForecast = (startIndex until endIndex).mapNotNull { i ->
        val epoch = parsedHourlyTimes[i]
        if (epoch < nowMillis) return@mapNotNull null // Extra safety

        HourlyForecast(
            time = epoch,
            temperature = Temperature.fromCelsius(hourly.temperature2m.getOrElse(i) { 0.0 }),
            weatherCode = hourly.weatherCode.getOrElse(i) { 0 },
            precipProbability = hourly.precipitationProbability.getOrElse(i) { 0 },
            windSpeed = WindSpeed.fromMetersPerSecond(hourly.windSpeed10m.getOrElse(i) { 0.0 }),
            windDirection = hourly.windDirection10m.getOrElse(i) { 0 },
            pressure = Pressure.fromHPa(hourly.pressureMsl.getOrElse(i) { 1013.0 }),
            precipitation = Precipitation.fromMm(hourly.precipitation.getOrElse(i) { 0.0 }),
        )
    }.distinctBy { it.time }

    val parsedDailyTimes = daily.time.map { parseIsoDate(it, utcOffsetSeconds) }

    val dailyForecast = daily.time.indices.map { i ->
        val epoch = parsedDailyTimes[i]
        DailyForecast(
            date = epoch,
            tempMax = Temperature.fromCelsius(daily.temperatureMax[i]),
            tempMin = Temperature.fromCelsius(daily.temperatureMin[i]),
            weatherCode = if (i == 0) c.weatherCode else daily.weatherCode[i],
            precipProbability = daily.precipitationProbabilityMax[i],
            windSpeed = WindSpeed.fromMetersPerSecond(daily.windSpeed10mMax.getOrElse(i) { 0.0 }),
            windDirection = daily.windDirection10mDominant.getOrElse(i) { 0 },
            precipitation = Precipitation.fromMm(daily.precipitationSum.getOrElse(i) { 0.0 }),
            uvIndexMax = daily.uvIndexMax.getOrElse(i) { 0.0 },
        )
    }.distinctBy { it.date }

    val minutelyForecastList = minutely15?.let { m15 ->
        val parsedM15Times = m15.time.map { parseIsoDateTime(it, utcOffsetSeconds) }
        val startM15Index = parsedM15Times.indexOfFirst { it >= nowMillis - 15 * 60 * 1000L }.takeIf { it != -1 } ?: 0
        val endM15Index = minOf(startM15Index + 12, m15.time.size)
        (startM15Index until endM15Index).map { i ->
            orinasa.njarasoa.maripanatokana.domain.model.MinutelyForecast(
                time = parsedM15Times[i],
                precipitation = Precipitation.fromMm(m15.precipitation.getOrElse(i) { 0.0 }),
            )
        }
    } ?: emptyList()

    return WeatherData(
        utcOffsetSeconds = utcOffsetSeconds,
        temperature = Temperature.fromCelsius(c.temperature),
        feelsLike = Temperature.fromCelsius(c.apparentTemperature),
        tempMin = Temperature.fromCelsius(daily.temperatureMin.firstOrNull() ?: c.temperature),
        tempMax = Temperature.fromCelsius(daily.temperatureMax.firstOrNull() ?: c.temperature),
        weatherCode = c.weatherCode,
        iconCode = wmoIconCode(c.weatherCode, isDay),
        locationName = locationName,
        locationSubtext = locationSubtext,
        pressure = Pressure.fromHPa(c.pressureMsl),
        humidity = c.relativeHumidity,
        dewPoint = Temperature.fromCelsius(c.dewPoint),
        windSpeed = WindSpeed.fromMetersPerSecond(c.windSpeed),
        windDeg = c.windDirection,
        windGust = if (c.windGusts > 0) WindSpeed.fromMetersPerSecond(c.windGusts) else null,
        rain = if (c.rain > 0) Precipitation.fromMm(c.rain) else null,
        snow = if (c.snowfall > 0) Precipitation.fromMm(c.snowfall) else null,
        cloudCover = c.cloudCover,
        uvIndex = c.uvIndex,
        visibility = c.visibility.toInt(),
        sunrise = sunriseEpoch,
        sunset = sunsetEpoch,
        dailySunrise = dailySunriseMillis,
        dailySunset = dailySunsetMillis,
        hourlyForecast = hourlyForecast,
        minutelyForecast = minutelyForecastList,
        dailyForecast = dailyForecast,
        alerts = emptyList(),
    )
}

File: web/src/lib/api/openMeteoMapper.ts
import type { OpenMeteoResponse } from './openMeteoTypes';
import type { WeatherData, HourlyForecast, DailyForecast } from '../domain/weatherData';
import { Temperature } from '../domain/temperature';
import { Pressure } from '../domain/pressure';
import { WindSpeed } from '../domain/windSpeed';
import { Precipitation } from '../domain/precipitation';

export function parseIsoDateTime(iso: string, utcOffsetSeconds: number): number {
	const year = parseInt(iso.substring(0, 4), 10);
	const month = parseInt(iso.substring(5, 7), 10) - 1;
	const day = parseInt(iso.substring(8, 10), 10);
	const hour = parseInt(iso.substring(11, 13), 10);
	const minute = parseInt(iso.substring(14, 16), 10);
	return Date.UTC(year, month, day, hour, minute) - utcOffsetSeconds * 1000;
}

function parseIsoDate(iso: string, utcOffsetSeconds: number): number {
	const year = parseInt(iso.substring(0, 4), 10);
	const month = parseInt(iso.substring(5, 7), 10) - 1;
	const day = parseInt(iso.substring(8, 10), 10);
	return Date.UTC(year, month, day) - utcOffsetSeconds * 1000;
}

export function mapToWeatherData(response: OpenMeteoResponse, locationName: string, locationSubtext?: string): WeatherData {
	const c = response.current;
	const d = response.daily;
	const h = response.hourly;

	const latitude = response.latitude;
	const longitude = response.longitude;

	const sunriseEpochSec = d.sunrise[0] ? Math.floor(parseIsoDateTime(d.sunrise[0], response.utc_offset_seconds) / 1000) : 0;
	const sunsetEpochSec = d.sunset[0] ? Math.floor(parseIsoDateTime(d.sunset[0], response.utc_offset_seconds) / 1000) : 0;

	const dailySunriseMillis = d.sunrise.map(t => parseIsoDateTime(t, response.utc_offset_seconds));
	const dailySunsetMillis = d.sunset.map(t => parseIsoDateTime(t, response.utc_offset_seconds));

	const nowMillis = Date.now();

	const parsedHourlyTimes = h.time.map(t => parseIsoDateTime(t, response.utc_offset_seconds));

	const startIndex = parsedHourlyTimes.findIndex((t) => t >= nowMillis);
	const hourlyForecast: HourlyForecast[] =
		startIndex === -1
			? []
			: h.time.slice(startIndex, startIndex + 24).map((time, i) => {
					const actualIndex = startIndex + i;
					const epoch = parsedHourlyTimes[actualIndex];
					return {
						time: epoch,
						temperature: Temperature.fromCelsius(h.temperature_2m[actualIndex]),
						weatherCode: h.weather_code[actualIndex],
						precipProbability: h.precipitation_probability[actualIndex],
						windSpeed: WindSpeed.fromMetersPerSecond(h.wind_speed_10m[actualIndex]),
						windDeg: h.wind_direction_10m[actualIndex],
						pressure: Pressure.fromHPa(h.pressure_msl[actualIndex]),
						precipitation: Precipitation.fromMm(h.precipitation[actualIndex]),
					};
				});

	const parsedDailyTimes = d.time.map(t => parseIsoDate(t, response.utc_offset_seconds));

	const dailyForecast: DailyForecast[] = d.time.map((time, i) => {
		return {
			date: parsedDailyTimes[i],
			tempMax: Temperature.fromCelsius(d.temperature_2m_max[i]),
			tempMin: Temperature.fromCelsius(d.temperature_2m_min[i]),
			weatherCode: i === 0 ? c.weather_code : d.weather_code[i],
			precipProbability: d.precipitation_probability_max[i],
			windSpeed: WindSpeed.fromMetersPerSecond(d.wind_speed_10m_max[i]),
			windDeg: d.wind_direction_10m_dominant[i],
			precipitation: Precipitation.fromMm(d.precipitation_sum[i]),
			uvIndexMax: d.uv_index_max?.[i] ?? 0,
		};
	});

	const m15 = response.minutely_15;
	const parsedM15Times = m15 ? m15.time.map((t) => parseIsoDateTime(t, response.utc_offset_seconds)) : [];
	const startM15Index = parsedM15Times.findIndex((t) => t >= nowMillis - 15 * 60 * 1000);
	const minutelyForecast = (m15 && startM15Index !== -1)
		? m15.time.slice(startM15Index, startM15Index + 12).map((_, i) => {
				const actualIndex = startM15Index + i;
				return {
					time: parsedM15Times[actualIndex],
					precipitation: Precipitation.fromMm(m15.precipitation[actualIndex] ?? 0),
				};
			})
		: [];

	return {
		utcOffsetSeconds: response.utc_offset_seconds,
		temperature: Temperature.fromCelsius(c.temperature_2m),
		feelsLike: Temperature.fromCelsius(c.apparent_temperature),
		tempMin: Temperature.fromCelsius(d.temperature_2m_min[0] ?? c.temperature_2m),
		tempMax: Temperature.fromCelsius(d.temperature_2m_max[0] ?? c.temperature_2m),
		weatherCode: c.weather_code,
		locationName,
		locationSubtext,
		latitude,
		longitude,
		pressure: Pressure.fromHPa(c.pressure_msl),
		humidity: c.relative_humidity_2m,
		dewPoint: Temperature.fromCelsius(c.dew_point_2m),
		windSpeed: WindSpeed.fromMetersPerSecond(c.wind_speed_10m),
		windDeg: c.wind_direction_10m,
		windGust: c.wind_gusts_10m > 0 ? WindSpeed.fromMetersPerSecond(c.wind_gusts_10m) : null,
		rain: c.rain > 0 ? Precipitation.fromMm(c.rain) : null,
		snow: c.snowfall > 0 ? Precipitation.fromMm(c.snowfall) : null,
		uvIndex: c.uv_index,
		cloudCover: c.cloud_cover,
		visibility: Math.round(c.visibility),
		sunrise: sunriseEpochSec,
		sunset: sunsetEpochSec,
		dailySunrise: dailySunriseMillis,
		dailySunset: dailySunsetMillis,
		hourlyForecast,
		minutelyForecast,
		dailyForecast,
		alerts: [],
		alertsLoading: true,
		timestamp: Date.now(),
	};
}

